src/widget/file_navigator/directory_view.rs

```diff
@@ -196,48 +197,52 @@ impl<'a> Widget for DirectoryView<'a> {
             });
 
             let show_hidden = self.show_hidden;
-            let entries: Vec<_> = match std::fs::read_dir(directory).ok() {
+            let mut entries: Vec<_> = match std::fs::read_dir(directory).ok() {
                 Some(entries) => {
                     entries.filter_map(|e| e.ok())
-                        .filter(|f| check_hidden(show_hidden, &f.path()))
-                        .collect()
+                        .filter_map(|f| {
+                            let path = f.path();
+                            if check_hidden(show_hidden, &path) {
+                                match types {
+                                    super::Types::All => Some(path),
+                                    super::Types::WithExtension(valid_exts) => {
+                                        // We're only after files.
+                                        if path.is_dir() {
+                                            Some(path)
+                                        } else {
+                                            // Check for valid extensions.
+                                            let ext = path.extension()
+                                                .and_then(|ext| ext.to_str())
+                                                .map(|s| std::ascii::AsciiExt::to_ascii_lowercase(s))
+                                                .unwrap_or_else(String::new);
+                                            if valid_exts.iter().any(|&valid_ext| &ext == valid_ext) {
+                                                Some(path)
+                                            } else {
+                                                None
+                                            }
+                                        }
+                                    },
+                                }
+                            } else {
+                                None
+                            }
+                        }).collect()
                 }
```

Comment:
I've pushed a version that moves all the filtering into check_hidden, avoiding all the right drift and making the code a little simpler to read.